### ISSUE TITLE
Improve the Virtualization docs to clarify that placeholder content must be the same size as item content

### ISSUE THREAD

**Issue by Alerinos** [dotnet/aspnetcore#44430]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Describe the bug

I used this example from the documentation:
https://learn.microsoft.com/pl-pl/aspnet/core/blazor/components/virtualization?view=aspnetcore-7.0
```cs
private async ValueTask<ItemsProviderResult<Employee>> LoadEmployees(
    ItemsProviderRequest request)
{
    var numEmployees = Math.Min(request.Count, totalEmployees - request.StartIndex);
    var employees = await EmployeesService.GetEmployeesAsync(request.StartIndex, 
        numEmployees, request.CancellationToken);

    return new ItemsProviderResult<Employee>(employees, totalEmployees);
}
```
My code looks like this:
```cs
    private async ValueTask<ItemsProviderResult<DTO.Article>> LoadEmployees(ItemsProviderRequest request)
    {
// Count = 336
        var count = await _search._context.Article.CountAsync(x => x.Status == Models.Article.StatusType.Show);
        var numEmployees = Math.Min(request.Count, count - request.StartIndex);
        
        var dto2 = await _search._context.Article
            .Include(x => x.Details)
            .Include(x => x.Names)
            .Include(x => x.Images)
            .Where(x => x.Status == Models.Article.StatusType.Show)
            .OrderByDescending(x => x.Details.Created)
            .Skip(request.StartIndex)
            .Take(numEmployees)
            .ToListAsync();

        var dto3 = dto2.Select(x => new DTO.Article
        {
            Name = x.Names.OrderByDescending(x => x.Created).Select(x => x.Name).FirstOrDefault() ?? string.Empty,
            Title = x.Details.Title,
            Description = x.Details.Description,
            Created = x.Details.Created,
            Tags = x.HashTags.Select(x => x.Tag).Take(3).ToList(),
            ImageHead = x.Images.LastOrDefault(x => x.Type == Image.StatusType.Header).Url ?? string.Empty,
        }).ToList();
        
// dto3 = 35 object
// count = 336 int
        return new ItemsProviderResult<DTO.Article>(dto3, count);
    }
```
My model:
```cs
internal class Article
{
    public Guid Id { get; set; }
    public string Status { get; set; } = string.Empty;
    public string Culture { get; set; } = string.Empty;
    public List<string> Cultures { get; set; } = null!;
    public string Name { get; set; } = string.Empty;
    public string Title { get; set; } = string.Empty;
    public string Content { get; set; } = string.Empty;
    public string Description { get; set; } = string.Empty;
    public bool Nsfw { get; set; } = false;
    public bool Sponsor { get; set; } = false;
    public bool Logged { get; set; } = false;
    public bool Premium { get; set; } = false;
    public decimal ReadingTime { get; set; }
    public DateTime Created { get; set; } = DateTime.UtcNow;
    public DateTime Updated { get; set; } = DateTime.UtcNow;
    public List<string> Tags { get; set; } = new();
    public string ImageHead { get; set; } = string.Empty;
    public List<Article> Articles { get; set; } = null!;
}
```
Razor:
```razor
<Row Class="gap-y-[1rem] gap-x-[1%]">

    <Virtualize Context="article" ItemsProvider="@LoadEmployees">
        <ItemContent>
            <Column Class="basis-full">
                <Card Title="@article.Title" Link="@($"Article/{article.Name}")" Hover=true>
                    <CardPicture Url="@article.ImageHead" Class="h-[15rem] xl:h-[20rem]" Sizes=@(new List<CardPicture.Size>{ new () { Width = 512, Height = 288 }})/>
                    <CardTitle Class="text-left text-xl"/>
                    <CardContent>
                        <HashTag Tag="@article.Tags" Class="inline"/>
                    <p class="inline text-md text-date dark:text-date_dark float-right">@article.Created.ToString("dd/MM/yy")</p>
                    </CardContent>   
                </Card>
            </Column>
        </ItemContent>
        <Placeholder>
            <Card>
                <CardTitle Title="Loading..."/>
                <CardContent>
                    In progres...
                </CardContent>
            </Card>
        </Placeholder>
    </Virtualize>

</Row>
```

The first results are correct. The problem starts with scrolling.

First loading, its ok:
![image](https://user-images.githubusercontent.com/42449535/194731707-c20af223-965c-4902-a850-f32f5d1513db.png)

1 break point scrool:
![image](https://user-images.githubusercontent.com/42449535/194731718-7b617779-4658-4386-a505-df37f044a71d.png)

The next one is right away, I don't have to do anything.
![image](https://user-images.githubusercontent.com/42449535/194731726-fdfe260f-ab06-4046-91c6-5361a7fc9ddf.png)

Scrolls on, 2 break point scroll:
![image](https://user-images.githubusercontent.com/42449535/194731740-e82dae9b-b13e-40fb-950e-3fb2b89280bf.png)

Scrolls on, 3 brak point scroll:
![image](https://user-images.githubusercontent.com/42449535/194731746-f74e80c8-ecc5-4bdf-ba1b-e66f829a18b9.png)

Still ok, now takes the breakpoint off.

```
[02:07:20 ERR] An exception occurred while iterating over the results of a query for context type 'Stand.Plugins.Articles.Context'.
System.InvalidOperationException: A second operation was started on this context instance before a previous operation completed. This is usually caused by different threads concurrently using the same instance of DbContext. For more information on how to avoid threading issues with DbContext, see https://go.microsoft.com/fwlink/?linkid=2097913.
   at Microsoft.EntityFrameworkCore.Infrastructure.Internal.ConcurrencyDetector.EnterCriticalSection()
   at Microsoft.EntityFrameworkCore.Query.Internal.SingleQueryingEnumerable`1.AsyncEnumerator.MoveNextAsync()
System.InvalidOperationException: A second operation was started on this context instance before a previous operation completed. This is usually caused by different threads concurrently using the same instance of DbContext. For more information on how to avoid threading issues with DbContext, see https://go.microsoft.com/fwlink/?linkid=2097913.
   at Microsoft.EntityFrameworkCore.Infrastructure.Internal.ConcurrencyDetector.EnterCriticalSection()
   at Microsoft.EntityFrameworkCore.Query.Internal.SingleQueryingEnumerable`1.AsyncEnumerator.MoveNextAsync()
[02:07:20 WRN] Unhandled exception rendering component: A second operation was started on this context instance before a previous operation completed. This is usually caused by different threads concurrently using the same instance of DbContext. For more information on how to avoid threading issues with DbContext, see https://go.microsoft.com/fwlink/?linkid=2097913.
System.InvalidOperationException: A second operation was started on this context instance before a previous operation completed. This is usually caused by different threads concurrently using the same instance of DbContext. For more information on how to avoid threading issues with DbContext, see https://go.microsoft.com/fwlink/?linkid=2097913.
   at Microsoft.AspNetCore.Components.Web.Virtualization.Virtualize`1.BuildRenderTree(RenderTreeBuilder builder)
   at Microsoft.AspNetCore.Components.Rendering.ComponentState.RenderIntoBatch(RenderBatchBuilder batchBuilder, RenderFragment renderFragment, Exception& renderFragmentException)
[02:07:20 ERR] Unhandled exception in circuit '7S3gOiTMV1OjU6v35dYZcg1i2Sb-MZExeXvGsHDWF2Q'.
System.InvalidOperationException: A second operation was started on this context instance before a previous operation completed. This is usually caused by different threads concurrently using the same instance of DbContext. For more information on how to avoid threading issues with DbContext, see https://go.microsoft.com/fwlink/?linkid=2097913.
   at Microsoft.AspNetCore.Components.Web.Virtualization.Virtualize`1.BuildRenderTree(RenderTreeBuilder builder)
   at Microsoft.AspNetCore.Components.Rendering.ComponentState.RenderIntoBatch(RenderBatchBuilder batchBuilder, RenderFragment renderFragment, Exception& renderFragmentException)
```

It turns out that everything is happening fast ...
I'm sitting for this half day today. I do not understand what's going on...
There is a documentation error or I don't understand how to do this.
Can I ask for help?

### .NET Version

.net 7 core rc1

**Comment by javiercn** [comment]
@Alerinos thanks for contacting us.

When you are scrolling fast, `LoadEmployees` is being called multiple times without the calls being properly serialized, which is causing multiple EF queries to be sent in parallel to the database (hence your error).

You have a cancellation token in the ItemsProviderRequest that you can flow to stop the ongoing queries before starting a new one.

**Comment by Alerinos** [comment]
@javiercn 
Scrolling slowly also causes a problem. I used something like that but it doesn't work:
```cs
      var count = await _search._context.Article.AsNoTracking().Where(x => x.Status == Models.Article.StatusType.Show).CountAsync(request.CancellationToken);  <--- here
        var numEmployees = Math.Min(request.Count, count - request.StartIndex);
        
        var dto2 = await _search._context.Article
            .AsNoTracking()
            .Include(x => x.Details)
            .Include(x => x.Names)
            .Include(x => x.Images)
            .Where(x => x.Status == Models.Article.StatusType.Show)
            .OrderByDescending(x => x.Details.Created)
            .Skip(request.StartIndex)
            .Take(request.Count)
            .ToListAsync(request.CancellationToken); <--- here
```
Are you able to see how it should be done? I think it is also worth updating the documentation, there may be many people like me.

**Comment by javiercn** [comment]
@Alerinos thanks for the additional details.

We already have documentation covering this. You can check it [here](https://learn.microsoft.com/en-us/aspnet/core/blazor/blazor-server-ef-core?view=aspnetcore-6.0). The suggestion would be to instantiate a new DbContext per operation as described in the article.

I am not 100% sure why the cancellation token approach did not work. That would be a question for the EF folks.

**Comment by Alerinos** [comment]
I found a similar problem.
https://github.com/dotnet/efcore/issues/26884
Unfortunately, nothing is explained in it, also marking it as a duplicate does not say anything.


@javiercn 
https://gyazo.com/ad72b054b5605326dd15a83cb2cf6df2
I think it is also Blazor's fault, see for yourself that instead of loading a new element once, it refreshes it many times until the exception.

my code:
```cs
    bool IsRun = false;

    private async ValueTask<ItemsProviderResult<DTO.Article>> LoadEmployees(ItemsProviderRequest request)
    {
        if (IsRun is true)
            await Task.Delay(TimeSpan.FromSeconds(1));

        IsRun = true;

...

        IsRun = false;

        return new ItemsProviderResult<DTO.Article>(dto3, count);
```

**Comment by javiercn** [comment]
@Alerinos to be more concrete, use a `DbContextFactory` to create a new DbContext inside `LoadEmployees`. I thought the documentation above mentioned that (One context per operation).

**Comment by Alerinos** [comment]
In the case of Cancel Token:
```
[12:27:12 ERR] An error occurred using the connection to database 'Stand' on server '(localdb)\mssqllocaldb'.
[12:27:12 WRN] Unhandled exception rendering component: A task was canceled.
System.Threading.Tasks.TaskCanceledException: A task was canceled.
   at Microsoft.AspNetCore.Components.Web.Virtualization.Virtualize`1.BuildRenderTree(RenderTreeBuilder builder)
   at Microsoft.AspNetCore.Components.Rendering.ComponentState.RenderIntoBatch(RenderBatchBuilder batchBuilder, RenderFragment renderFragment, Exception& renderFragmentException)
[12:27:12 ERR] Unhandled exception in circuit '-hj7HmI-Jd_Mt8oxwAQwPL_s3B8LSC4EDqUgaZb06xI'.
System.Threading.Tasks.TaskCanceledException: A task was canceled.
   at Microsoft.AspNetCore.Components.Web.Virtualization.Virtualize`1.BuildRenderTree(RenderTreeBuilder builder)
   at Microsoft.AspNetCore.Components.Rendering.ComponentState.RenderIntoBatch(RenderBatchBuilder batchBuilder, RenderFragment renderFragment, Exception& renderFragmentException)
```

Without:
https://user-images.githubusercontent.com/42449535/194846863-ffa547a3-b8e6-41a1-b97b-693d82127f1a.mp4

```cs

    [Inject] IDbContextFactory<Context> DbFactory { get; set; } = null!;

    private async ValueTask<ItemsProviderResult<DTO.Article>> LoadEmployees(ItemsProviderRequest request)
    {
        using var context = DbFactory.CreateDbContext();
        
        var count = await context.Article.AsNoTracking().Where(x => x.Status == Models.Article.StatusType.Show).CountAsync(); // request.CancellationToken
        var numEmployees = Math.Min(request.Count, count - request.StartIndex);
        
        var dto2 = await context.Article
            .AsNoTracking()
            .Include(x => x.Details)
            .Include(x => x.Names)
            .Include(x => x.Images)
            .Where(x => x.Status == Models.Article.StatusType.Show)
            .OrderByDescending(x => x.Details.Created)
            .Skip(request.StartIndex)
            .Take(request.Count)
            .ToListAsync(); // request.CancellationToken
```

@javiercn I understand that this is correct behavior and is it supposed to be?

**Comment by javiercn** [comment]
@Alerinos it is not obvious what is going on in your app.

Could you provide a minimal repro project so that we can try and determine if there is an issue here? 

At this point I do not think I have the expertise in this area to be accurate, so someone more familiar with the code will help here.

**Comment by Alerinos** [comment]
@javiercn 
[Virtualize-bug.zip](https://github.com/dotnet/aspnetcore/files/9745991/Virtualize-bug.zip)
Just run, the database will create itself.

**Comment by javiercn** [comment]
@Alerinos I think this will likely come down to an issue in the sample app (once we get a chance to look at it) or a more complete doc, I have peek through the framework code, and I don't see anything obviously wrong.

**Comment by Alerinos** [comment]
@javiercn 
```cs
 await Task.Delay(TimeSpan.FromSeconds(1));
```

I used something like that. It seems that the component does not wait for return to be executed, but creates a new instance.
I don't think it works properly asynchronously.

I will try to test with rest api.
<details>
  <summary>A whole class of site</summary>

```cs
using Microsoft.AspNetCore.Components.Web.Virtualization;
using Microsoft.AspNetCore.Components;
using Microsoft.EntityFrameworkCore;

namespace Virtualize_bug.Pages;

[Route("Articles")]
public partial class Articles : ComponentBase
{
    [Inject] IDbContextFactory<Context> DbFactory { get; set; } = null!;

    private async ValueTask<ItemsProviderResult<DTO.Article>> LoadEmployees(ItemsProviderRequest request)
    {
        await Task.Delay(TimeSpan.FromSeconds(1));

        using var context = DbFactory.CreateDbContext();

        var count = await context.Article.AsNoTracking().CountAsync(); // request.CancellationToken
        var numEmployees = Math.Min(request.Count, count - request.StartIndex);

        var dto2 = await context.Article
            .AsNoTracking()
            .Include(x => x.Details)
            .OrderByDescending(x => x.Details.Created)
            .Skip(request.StartIndex)
            .Take(request.Count)
            .ToListAsync(); // request.CancellationToken

        var dto3 = dto2.Select(x => new DTO.Article
        {
            Title = x.Details.Title,
            Description = x.Details.Description,
            Created = x.Details.Created
        }).ToList();

        return new ItemsProviderResult<DTO.Article>(dto3, count);
    }
}
```
</details>

**Comment by mkArtakMSFT** [comment]
@Alerinos can you please share a minimum repro project hosted as a public GitHub repo, so that we can investigate this further. Thanks!

**Comment by Alerinos** [comment]
@mkArtakMSFT https://github.com/Alerinos/Virtualize-bug

**Comment by MackinnonBuck** [comment]
Thanks for the repro project, @Alerinos!

The `Virtualize` component requires that all content items have identical height, and this includes the placeholder content. After changing the repro project that you provided to have identically sized items, the scrolling weirdness demonstrated in your video goes away.

Could you confirm whether applying this fix to your own project solves the issue? Thanks!

**Comment by Alerinos** [comment]
@MackinnonBuck When it comes to adding height in css (like below) there is still a problem.

https://github.com/Alerinos/Virtualize-bug/blob/main/Pages/Articles.razor#L3

**Comment by MackinnonBuck** [comment]
@Alerinos, could you apply that same styling to the `<Placeholder>` as well? For example:

```razor
﻿<Virtualize Context="article" ItemsProvider="@LoadEmployees">
    <ItemContent>
        <div style="height: 50px;">
            <h3>@article.Title</h3>
            <p>@article.Content</p>
        </div>
    </ItemContent>
    <Placeholder>
        <div style="height: 50px;">
            In progress...
        </div>
    </Placeholder>
</Virtualize>
```

**Comment by Alerinos** [comment]
@MackinnonBuck 
Seems to be working, thanks for your help. You need to complete the documentation. I will test in production and let me know if there are any problems.

As we are talking, I have another problem, I want to change the URL while loading. (In the case of refreshing the page, the user will return to the element he saw, not from the beginning)
Unfortunately, I have this problem:
https://gyazo.com/b5a24b55793155acb725c4b9755931ab

https://github.com/Alerinos/Virtualize-bug/blob/main/Pages/Articles.razor.cs#L41

Is it possible to work around this problem or create an event that allows this behavior?

**Comment by MackinnonBuck** [comment]
@Alerinos sorry, but it's hard to tell what the issue is in that video. What's going wrong, exactly?

> In the case of refreshing the page, the user will return to the element he saw, not from the beginning

This is something we don't support yet, but there is an issue tracking it: https://github.com/dotnet/aspnetcore/issues/26943

**Comment by Alerinos** [comment]
@MackinnonBuck 
In fact, there is no problem in the movie. The point is that when an event is triggered:
```
NavigationManager.NavigateTo($"/Articles/?s={request.StartIndex}", replace: true);
```
The page goes up to the top instead of accepting the URL and staying put.
I don't know if this is a NavigationManager or Virtual component problem

You can see the problem here. The page goes up by the way, every time the url changes again so how would the component work and try to change the value.
https://gyazo.com/99132fdf2fd459b6442dd50287c95f5a

**Comment by MackinnonBuck** [comment]
> The page goes up to the top instead of accepting the URL and staying put.

Yeah, unfortunately `NavigationManager` does this and currently there isn't API to disable that behavior. But you can work around it today by doing something like this: https://github.com/dotnet/aspnetcore/issues/40190#issuecomment-1203857906

Also, regarding:

> You need to complete the documentation.

The [docs](https://learn.microsoft.com/en-us/aspnet/core/blazor/components/virtualization?view=aspnetcore-7.0#advanced-styles-and-scroll-detection-1) _do_ state:

> Virtualize works under the following conditions:

> * All content items are of identical height. This makes it possible to calculate which content corresponds to a given scroll position without first fetching every data item and rendering the data into a DOM element.
...

However, the phrase "content item" isn't really defined in that document, and it's not made explicit elsewhere that `<Placeholder>` content is included in the "all items must have identical height" constraint. So, I agree the docs can be improved.

Since there isn't a bug in the framework, let's use this issue to track improving the wording of the docs to make the placeholder content requirements more obvious.

**Comment by Alerinos** [comment]
Thank you very much for your help.

@MackinnonBuck  I see one more potential problem, I need to test it.
What if the application uses RWD? The page on the computer looks different than on the mobile device? Example:

![image](https://user-images.githubusercontent.com/42449535/199623910-0b15edef-c941-4ba8-86a7-1640648f0e08.png)

Once, when I tested (a year ago) there was a problem with it, then I let it go.

**Comment by MackinnonBuck** [comment]
@Alerinos, we have an issue tracking that as well: https://github.com/dotnet/aspnetcore/issues/28821

**Comment by Alerinos** [comment]
@MackinnonBuck 
Now I think so, there should be some protection against a large amount of reordering in a virtual component. What if someone erases the height in the browser (or the browser does it itself)? There will be a large number of requests to the server and ddos can be created.
It is worth discussing this topic and thinking about some sensible safeguard. Maybe the number of requests within e.g. 1 second?

**Comment by MackinnonBuck** [comment]
@Alerinos That seems like a reasonable suggestion. We currently have https://github.com/dotnet/aspnetcore/issues/10522 tracking the general idea of throttling/debouncing events, but the `Virtualize` case might require special attention. It doesn't look like we have an issue for throttling `Virtualize` specifically, so feel free to open a separate issue for it 🙂 Thanks!

**Comment by Alerinos** [comment]
Thanks, I created a new topic.

@MackinnonBuck  I found another problem, the component does not "cache" HTML elements but cleans them, so there may be another security problem. See in the attached video how I load the database by playing only with the scroll. You should also consider not disposing of the loaded item or keeping it somewhere in memory.

https://gyazo.com/1cf98d0ad8c29323ba843bd790e0725a

**Comment by MackinnonBuck** [comment]
Thanks, @Alerinos.

> You should also consider not disposing of the loaded item or keeping it somewhere in memory.

This can already be done by manually caching results and returning them from the `ItemsProvider`.

**Comment by Alerinos** [comment]
> This can already be done by manually caching results and returning them from the ItemsProvider.

I meant more on the browser side, thanks to which we will save transfer between the server and the client. For large pages, this will make a big difference.

**Comment by MackinnonBuck** [comment]
@Alerinos If I'm understanding correctly, you're suggesting we cache the removed DOM elements on the JavaScript side and reinsert them (still via JavaScript) when they get scrolled back into view? I'm inclined to think that's not a direction we would take for the following reasons:

* This only works if the content in the list never changes beyond the first render. If we were to instead modify this suggestion to only use the previously rendered elements as placeholders while waiting for the server to send updates, it would make the UI slightly more responsive, but it wouldn't reduce network bandwidth.
* Implementing this would require that we duplicate a lot of the .NET virtualization logic on the JavaScript side and make the two implementations compatible.
* There would likely be a lot of complexity around having the DOM mutations dictated by the new JavaScript logic play nicely with the DOM updates sent by the server (generally speaking, you have to be really careful about having JavaScript mutate any part of the DOM generated by Blazor).

So, this might end up resulting in a lot of added complexity for marginal gain, but please correct me if I'm misinterpreting the suggestion 🙂

**Comment by Alerinos** [comment]
@MackinnonBuck 

To display displayed items:
![image](https://user-images.githubusercontent.com/42449535/200094179-0a66020f-639d-4c14-bf7d-f6b89919965b.png)
When moving, old elements are deleted and new ones are added:
![image](https://user-images.githubusercontent.com/42449535/200094218-4905ecf4-f64f-4160-a09f-410f8dcb0951.png)

Instead of deleting them, we can leave them and possibly give an ID so that blazor knows which elements are already loaded.
I think that we do not need to delete HTML that has already been downloaded

https://gyazo.com/bbed2cdd93a7adfe88721ec6516f3854

Another thing I would improve is:
```
return new ItemsProviderResult<DTO.Article>(dto3, count);
```
Return total quantity of items. I believe that it is unnecessary. Why?

We have to make an additional query to the server how many elements are there. It is inconvenient and resource-consuming.
How would I go about it?

If the first parameter is zero, it means that there are no elements, you can additionally handle this exception like this:

```razor
<Virtualize Context="article" ItemsProvider="@LoadEmployees">
    <ItemContent>
        <div style="height: 50px;">
            <h3>@article.Title</h3>
            <p>@article.Content</p>
        </div>
    </ItemContent>
    <Placeholder>
        <div style="height: 50px;">
            In progress...
        </div>
    </Placeholder>
    <Missing>
        <div style="height: 50px;">
           Sorry, no content, please continue or refresh the page.
        </div>
    </Missing>
</Virtualize>
```
I think that something like this would make the work of developers easier and save resources. Let me know what you think.

**Comment by Alerinos** [comment]
Another problem I see is when we have 200 articles. Someone adds 20 items while browsing, the cout will change. How will the component react? We will not have duplicate data? At the time of viewing the data, there were fewer items to display than there is now, the new items are higher.

**Comment by Alerinos** [comment]
I did another test.

If we have a high screen, for example, we are shown 30 records, each downward movement causes reloading of 30 elements. (Why load the same if it already exists)

What's the problem? The same data is downloaded every time you move up or down.

How should it be done in my opinion?
Each move should charge +3 items, i.e. moving down will charge, +3 then +3 etc.

**Comment by MackinnonBuck** [comment]
> I think that we do not need to delete HTML that has already been downloaded

If you want to preserve already-rendered HTML, then `Virtualize` probably isn't the component you're looking for. The `Virtualize` component was designed to be capable of handling quantities of items that would be too large to reasonably render all at once on the page without performance problems. If this capability isn't a requirement for your use case, and if frequent communication with the server caused by scrolling is a legitimate concern, then it's probably better to use a custom component.

> Return total quantity of items. I believe that it is unnecessary. Why?

It's necessary because `Virtualize` needs to calculate how large the scrollable area should be, and it needs the total number of items to do so. If you can't compute the actual total number of items, you can specify it as (_number of currently-loaded items_ + 1) and you'll get a sort of infinite scroll effect where the size of the scrollable area expands as more items are loaded.

> If the first parameter is zero, it means that there are no elements, you can additionally handle this exception like this...

https://github.com/dotnet/aspnetcore/issues/28770

---

It seems you have a lot of ideas and concerns, and that's awesome. However, it's best that we keep comments on this issue related to this issue's topic, which we've decided should be improving the `Virtualize` docs to make the "all item content should be the same size" requirement clearer.

Going forward, would you please use the following process when making feature requests?
1. See if there is already an issue logged for the feature you're asking about, and if there is, upvote it so it shows up as higher on the list for our planning process.
2. If there is not an issue tracking the feature you want addressed, file a new one. This way, we can track and address each item separately.

Thanks for your feedback and enthusiasm!

**Comment by szalapski** [comment]
>It's necessary because Virtualize needs to calculate how large the scrollable area should be, and it needs the total number of items to do so. If you can't compute the actual total number of items, you can specify it as (number of currently-loaded items + 1) and you'll get a sort of infinite scroll effect where the size of the scrollable area expands as more items are loaded.

I like this idea.  @MackinnonBuck , are you saying that this is possible now?  I don't see a property on Virtualize with the number of items to keep in DOM, nor do I see a way to get the number of items currently in the DOM.

**Comment by MackinnonBuck** [comment]
@szalapski Sorry for the late reply - just saw your question.

Here's a basic demonstration of how this could work:
```razor
<Virtualize ItemsProvider="LoadItemsAsync">
    <ItemContent>
        <p @key="context">@context</p>
    </ItemContent>
</Virtualize>

@code {
    private const int PageSize = 40;

    private int totalItemCount = 20;

    private async ValueTask<ItemsProviderResult<int>> LoadItemsAsync(ItemsProviderRequest request)
    {
        var items = Enumerable.Range(request.StartIndex, request.Count);
        if (request.StartIndex + request.Count >= totalItemCount)
        {
            // Simulate an artificial delay when loading new content
            await Task.Delay(250, request.CancellationToken);
            totalItemCount += PageSize;
        }

        return new(items, totalItemCount);
    }
}
```

**Comment by MackinnonBuck** [comment]
@guardrex Would you be able to clarify the virtualization docs, as suggested in [this comment](https://github.com/dotnet/aspnetcore/issues/44430#issuecomment-1301500518)?

(I know there's a lot going on in this issue, but we're just using it to track improving this area of the docs)

**Comment by guardrex** [comment]
Addressed on https://github.com/dotnet/AspNetCore.Docs/pull/31499.

**Comment by MackinnonBuck** [comment]
Thanks @guardrex! Closing this out.
